cmd/kubelet/app/server.go

```diff
@@ -1173,19 +1174,26 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
 		switch v1.ResourceName(k) {
 		// CPU, memory, local storage, and PID resources are supported.
 		case v1.ResourceCPU, v1.ResourceMemory, v1.ResourceEphemeralStorage, pidlimit.PIDs:
-			if v1.ResourceName(k) != pidlimit.PIDs || utilfeature.DefaultFeatureGate.Enabled(features.SupportNodePidsLimit) {
-				q, err := resource.ParseQuantity(v)
-				if err != nil {
-					return nil, err
-				}
-				if q.Sign() == -1 {
-					return nil, fmt.Errorf("resource quantity for %q cannot be negative: %v", k, v)
-				}
-				rl[v1.ResourceName(k)] = q
-			}
 		default:
+			// hugepages are also supported
+			if v1helper.IsHugePageResourceName(v1.ResourceName(k)) {
+				break
+			}
 			return nil, fmt.Errorf("cannot reserve %q resource", k)
 		}
+
+		if v1.ResourceName(k) == pidlimit.PIDs && !utilfeature.DefaultFeatureGate.Enabled(features.SupportNodePidsLimit) {
+			return nil, fmt.Errorf("cannot reserve %q resource", k)
+		}
+
+		q, err := resource.ParseQuantity(v)
+		if err != nil {
+			return nil, err
+		}
+		if q.Sign() == -1 {
+			return nil, fmt.Errorf("resource quantity for %q cannot be negative: %v", k, v)
+		}
+		rl[v1.ResourceName(k)] = q
```

Comment:
One thing I'm not seeing is code where we actually "reserve" the hugepages... but am I correct that adding the hugepage resources to the `r1` map is sufficient for them to be reserved? I.e. the function which takes `r1` and reserves its contents can handle hugepages already?

Reply:
It should happen in here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/node_container_manager_linux.go#L57

Need to verify that there is no reason it wouldn't just get picked up.

Reply:
And you can see huge page limit is pulled here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/node_container_manager_linux.go#L165

Reply:
@derekwaynecarr is correct.

This function is used here: https://github.com/kubernetes/kubernetes/blob/866ea74/cmd/kubelet/app/server.go#L647-L654

I do agree that the naming (and the comment above) is kinda bad, so renaming it to `parseReservedResources` (or `parseSystemResources` would make more sense. I think it is outside the scope of this PR, but if you agree, I can file another one with the change.

Reply:
Hi @bg-chun,

> 1. Allocatable of Node Allocatable feature indicates the amount of compute resources that are available for pods.
>2. The Scheduler(master component) will take Allocatable resources from multiple nodes to schedule a Pod to a Node.
>3. At this point, Kubelet treats entire pre-allocated hugepages as allocatable hugepages of a node.
>4. by reserving hugepages by system(and kube)-reserved, we can "exclude" a certain amount of hugepages from the pool of allocatable hugepages.

1. Correct
2. Correct
3. Correct, kubelet will add all available pages into the Capacity of the nodes, and `Allocatable = (capacity) - reserved`
4. Correct. We do however still talk about "amount of huge page memory" and not a discrete page count. So you can eg. reserve `2GiB` of `1GiB` hugepages for the system.

I think the docs will make it easier to understand how things works: https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/